core/src/main/java/org/elasticsearch/search/builder/NewSearchSourceBuilder.java

```diff
@@ -0,0 +1,1077 @@
+            builder.fieldDataFields = fieldDataFields;
+        }
+        boolean hasFieldNames = in.readBoolean();
+        if (hasFieldNames) {
+            int size = in.readVInt();
+            List<String> fieldNames = new ArrayList<>(size);
+            for (int i = 0; i < size; i++) {
+                fieldNames.add(in.readString());
+            }
+            builder.fieldNames = fieldNames;
+        }
+        builder.from = in.readVInt();
+        if (in.readBoolean()) {
+            builder.highlightBuilder = in.readBytesReference();
+        }
+        boolean hasIndexBoost = in.readBoolean();
+        if (hasIndexBoost) {
+            int size = in.readVInt();
+            ObjectFloatHashMap<String> indexBoost = new ObjectFloatHashMap<String>(size);
+            for (int i = 0; i < size; i++) {
+                indexBoost.put(in.readString(), in.readFloat());
+            }
+            builder.indexBoost = indexBoost;
+        }
+        if (in.readBoolean()) {
+            builder.innerHitsBuilder = in.readBytesReference();
+        }
+        if (in.readBoolean()) {
+            builder.minScore = in.readFloat();
+        }
+        if (in.readBoolean()) {
+            builder.postQueryBuilder = in.readQuery();
+        }
+        if (in.readBoolean()) {
+            builder.queryBuilder = in.readQuery();
+        }
+        if (in.readBoolean()) {
+            int size = in.readVInt();
+            List<BytesReference> rescoreBuilders = new ArrayList<>();
+            for (int i = 0; i < size; i++) {
+                rescoreBuilders.add(in.readBytesReference());
+            }
+            builder.rescoreBuilders = rescoreBuilders;
+        }
+        if (in.readBoolean()) {
+            int size = in.readVInt();
+            List<ScriptField> scriptFields = new ArrayList<>(size);
+            for (int i = 0; i < size; i++) {
+                scriptFields.add(ScriptField.PROTOTYPE.readFrom(in));
+            }
+            builder.scriptFields = scriptFields;
+        }
+        builder.size = in.readVInt();
+        if (in.readBoolean()) {
+            int size = in.readVInt();
+            List<BytesReference> sorts = new ArrayList<>();
+            for (int i = 0; i < size; i++) {
+                sorts.add(in.readBytesReference());
+            }
+            builder.sorts = sorts;
+        }
+        if (in.readBoolean()) {
+            builder.stats = in.readStringArray();
+        }
+        if (in.readBoolean()) {
+            builder.suggestBuilder = in.readBytesReference();
+        }
+        builder.terminateAfter = in.readVInt();
+        builder.timeoutInMillis = in.readVLong();
+        builder.trackScores = in.readBoolean();
+        builder.version = in.readBoolean();
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        boolean hasAggregations = aggregations != null;
+        out.writeBoolean(hasAggregations);
+        if (hasAggregations) {
```

Comment:
can we use a boolean for this and special case it? I'd for xcontent use `if (list.isEmpty() == false)` instead of the null check so we don't render it?